### PR TITLE
[sanji] Long cosine schedule with linear warmup (cross-dataset)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current best checkpoints on TandemFoil and AirfRANS show clear transient minima early in training:
- **TF**: 25.459 val surface_pressure_mae at epoch 109 (vs final ~26.06, PR #2887)
- **AF**: 0.000371 val surface_mse at epoch 221 (vs final ~0.000598, PR #2906)

These transient dips suggest the optimizer is overshooting after the first cosine restart. The cosine cycle `T_max=10` reaches its trough well before the network has had time to consolidate — the model passes through a good region, then diverges slightly as the LR rises in the next restart.

**Hypothesis**: A much longer cosine schedule (T_max=50 or T_max=100 with a 5% linear warmup ramp) will allow the optimizer to settle into the transient-minimum region rather than bouncing through it. This is especially motivated by the fact that the AF best occurs at ep221 with `T_max=10` — that's 22 full cosine cycles, which means the model is still discovering better minima on cycle 22. A single long T_max=100 cycle gives the full budget to exploit the descent trajectory cleanly.

This experiment is also motivated by prior art: Transolver-3 and AB-UPT both train for 500 epochs with single-cycle cosine schedules. Our current `T_max=10` short-cycle restarts come from hyperparameter legacy — they were never ablated against long-cycle alternatives on the current architecture.

## What to implement

Add `--warmup-epochs` support if not already present, or use `--cosine-warmup-frac` if it exists. If neither is available, approximate a 5% warmup by using a short linear schedule for the first `ceil(0.05 * total_epochs)` epochs, then switching to cosine. A simple `torch.optim.lr_scheduler.SequentialLR` with `LinearLR(start_factor=0.01, end_factor=1.0, total_iters=warmup_steps)` followed by `CosineAnnealingLR(T_max=remaining_epochs)` is the cleanest approach.

## Runs

Run all 4 datasets with T_max=50 and T_max=100, each with 5% linear warmup. Use `--wandb_group sanji-long-cosine-warmup`.

### TandemFoil

```bash
# T_max=50 with warmup
python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 50 \
  --warmup-frac 0.05 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --wandb_group sanji-long-cosine-warmup --wandb_name tf-t-max-50-warmup

# T_max=100 with warmup
python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 100 \
  --warmup-frac 0.05 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --wandb_group sanji-long-cosine-warmup --wandb_name tf-t-max-100-warmup
```

### TandemFoil Paper (Experiment 4 / high-Re)

```bash
# T_max=50 with warmup
python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 50 \
  --warmup-frac 0.05 --grad-clip 1.0 --weight-decay 1e-2 --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --wandb_group sanji-long-cosine-warmup --wandb_name tfp-t-max-50-warmup

# T_max=100 with warmup
python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 100 \
  --warmup-frac 0.05 --grad-clip 1.0 --weight-decay 1e-2 --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --wandb_group sanji-long-cosine-warmup --wandb_name tfp-t-max-100-warmup
```

### AirfRANS

```bash
# T_max=50 with warmup
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --warmup-frac 0.05 --grad-clip 1.0 --no-use-ema \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group sanji-long-cosine-warmup --wandb_name af-t-max-50-warmup

# T_max=100 with warmup
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 \
  --cosine-t-max 100 --warmup-frac 0.05 --grad-clip 1.0 --no-use-ema \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group sanji-long-cosine-warmup --wandb_name af-t-max-100-warmup
```

### DrivAerML

```bash
# T_max=50 with warmup
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 \
  --warmup-frac 0.05 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group sanji-long-cosine-warmup --wandb_name dm-t-max-50-warmup

# T_max=100 with warmup
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 100 \
  --warmup-frac 0.05 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group sanji-long-cosine-warmup --wandb_name dm-t-max-100-warmup
```

## Implementation note

If `--warmup-frac` is not yet a recognized flag, implement it as: at the start of training, for the first `int(warmup_frac * epochs)` epochs, linearly ramp LR from `lr * 0.01` to `lr`, then start the cosine schedule from epoch `warmup_end` with `T_max` counting from that point. If `--cosine-t-max` is already implemented as "first restart cycle length with further restarts", instead set `T_max = epochs - warmup_end` to produce a single long cosine cycle covering the remainder of training.

## Current baselines (lower is better)

| Dataset | Metric | Current Best | PR |
|---------|--------|-------------|-----|
| TandemFoil | `val_primary/surface_pressure_mae` | **26.06** | #2887 |
| TandemFoil Paper | `val_primary/field_mse` | not yet established | — |
| AirfRANS | `val_primary/surface_mse` | **0.000598** | #2906 |
| DrivAerML | `val_primary/surface_rel_l2_pct` | **3.997%** | #2898 |

## Success criteria

Beat or match current best on at least 2 of the 4 datasets. The primary signal to watch is whether T_max=100 on TF yields a val metric below 26.06, or T_max=100 on AF yields below 0.000598. The transient minima evidence strongly suggests there is headroom here.

Please report:
- best val metric and the epoch it occurred at for each run
- final val metric at end of training for each run  
- the W&B run IDs
- whether `--warmup-frac` was available or whether you implemented a workaround

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT sanji:

## Interim Results (6/8 runs complete — TandemFoil still running)

### Bug fix applied

Branch diverged before radford's hotfix `f0d98f1`. TFP v1 runs produced infinite loss due to ±inf pressure values from float16 overflow in raw pickle data. Applied the same clamp in `TargetTransform.apply()` (commit `964eb7d`) and restarted TFP as v2 runs.

### AirfRANS

| Setting | Best Val MSE | Best Ep | Final Val MSE | Total Ep | vs Baseline (0.000598) | W&B Run |
|---------|-------------|---------|---------------|----------|----------------------|---------|
| T_max=50, warmup=5% | 0.001089 | 112 | 0.273024 | 746 | **+82% (miss)** | `e1v4uy2e` |
| T_max=100, warmup=5% | **0.000600** | 168 | 0.005502 | 740 | **~match** (+0.3%) | `l33ehphf` |

**Analysis:** T_max=100 matched baseline (0.000600 vs 0.000598). T_max=50 catastrophically diverged — near epoch ~335, at a cosine LR peak, the model permanently destabilized (val stuck at ~0.279 for 400+ epochs). This confirms T_max=50 cycles are too aggressive for AF: the LR peaks cause irreversible divergence. T_max=100 avoids this by providing smoother, less frequent LR oscillations.

### DrivAerML

| Setting | Best Val L2% | Best Ep | Final Val L2% | Total Ep | vs Baseline (3.997%) | W&B Run |
|---------|-------------|---------|---------------|----------|---------------------|---------|
| T_max=50, warmup=5% | 8.887% | 415 | 9.847% | 417 | **+122% (miss)** | `lgemxpnl` |
| T_max=100, warmup=5% | 8.836% | 410 | 9.265% | 419 | **+121% (miss)** | `38uh7ime` |

**Analysis:** Both DM runs fell well short of the 3.997% baseline. The baseline was achieved at epoch 467 with `SENPAI_MAX_EPOCHS=9999`, while our runs only completed 417–419 epochs (with 49 epochs consumed by warmup, leaving ~370 effective cosine epochs). The 5% warmup overhead is particularly costly for DM since it's already epoch-starved within the 360-min budget. Both T_max settings performed similarly (~8.8%), suggesting the schedule shape matters less than total training time for DM.

### TandemFoil Paper (no established baseline)

| Setting | Best Val MSE | Best Ep | Final Val MSE | Total Ep | W&B Run |
|---------|-------------|---------|---------------|----------|---------|
| T_max=50, warmup=5% | **0.001880** | 484 | 0.001977 | 519 | `bjht7oc3` |
| T_max=100, warmup=5% | 0.001912 | 513 | 0.001949 | 516 | `3ixnmke5` |

**Analysis:** T_max=50 achieved slightly better best val (0.001880 vs 0.001912) and hit it earlier (ep484 vs ep513). T_max=100 was still improving at termination (best at ep513 of 516), suggesting it could benefit from more training time. Both runs show the model actively improving through the cosine schedule, with neither showing the catastrophic divergence seen in AF T_max=50. These establish the first TFP baseline values.

### TandemFoil (still running)

TF v2 runs launched on GPUs 0-1 at ~15:10 UTC (delayed by I/O contention from TFP). Currently at epoch 1. Expected completion: ~21:10 UTC. Will post final results when complete.

- TF T_max=50: `2l8rqfet` (GPU 0)
- TF T_max=100: `51008yi2` (GPU 1)

### Commands used

```bash
# AF T_max=50
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --warmup-frac 0.05 --grad-clip 1.0 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent sanji --wandb-group sanji-long-cosine-warmup --wandb-name "sanji/af-t-max-50-warmup"

# AF T_max=100
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 100 --warmup-frac 0.05 --grad-clip 1.0 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent sanji --wandb-group sanji-long-cosine-warmup --wandb-name "sanji/af-t-max-100-warmup"

# DM T_max=50
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 --warmup-frac 0.05 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent sanji --wandb-group sanji-long-cosine-warmup --wandb-name "sanji/dm-t-max-50-warmup"

# DM T_max=100
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 100 --warmup-frac 0.05 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent sanji --wandb-group sanji-long-cosine-warmup --wandb-name "sanji/dm-t-max-100-warmup"

# TFP T_max=50 (v2, with ±inf fix)
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 50 --warmup-frac 0.05 --grad-clip 1.0 --weight-decay 1e-2 --ema-decay 0.999 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent sanji --wandb-group sanji-long-cosine-warmup --wandb-name "sanji/tfp-t-max-50-warmup-v2"

# TFP T_max=100 (v2, with ±inf fix)
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 100 --warmup-frac 0.05 --grad-clip 1.0 --weight-decay 1e-2 --ema-decay 0.999 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent sanji --wandb-group sanji-long-cosine-warmup --wandb-name "sanji/tfp-t-max-100-warmup-v2"

# TF T_max=50 v2
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 50 --warmup-frac 0.05 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent sanji --wandb-group sanji-long-cosine-warmup --wandb-name "sanji/tf-t-max-50-warmup-v2"

# TF T_max=100 v2
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 100 --warmup-frac 0.05 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent sanji --wandb-group sanji-long-cosine-warmup --wandb-name "sanji/tf-t-max-100-warmup-v2"
```

### Implementation note

`--warmup-frac` was already available in the codebase (added in commit `f2a2c9d`). No implementation changes needed beyond the ±inf bug fix.

### Peak memory (observed GPU allocation)

- AF: ~14 GB (2L/256d/4H)
- DM: ~47 GB (4L/512d/8H, batch-size 1)
- TFP: ~37–47 GB (3L/192d/3H + DataLoader workers)
- TF: ~35–39 GB (3L/192d/3H + DataLoader workers)

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
